osutil/disks/mockdisk.go

```diff
@@ -0,0 +1,144 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package disks
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+)
+
+var (
+	isSnapdTest = len(os.Args) > 0 && strings.HasSuffix(os.Args[0], ".test")
+)
+
+func panicIfNotTest() {
+	// only to be used in tests!!!!
+	if !isSnapdTest {
+		panic("mocking functions only to be used in tests!")
+	}
+}
+
+// MockDiskMapping is an implementation of Disk for mocking purposes, it is
+// exported so that other packages can easily mock a specific disk layout
+// without needing to mock the mount setup, sysfs, or udev commands just to test
+// high level logic. We use an embedded type of an unexported type so as not to
+// allow arbitrary creation of MockDiskMapping.
+type MockDiskMapping struct {
+	FilesystemLabelToPartUUID map[string]string
+	DiskHasPartitions         bool
+}
+
+// FindMatchingPartitionUUID returns a matching PartitionUUID for the specified
+// label if it exists. Part of the Disk interface.
+func (d *MockDiskMapping) FindMatchingPartitionUUID(label string) (string, error) {
+	panicIfNotTest()
+	if partuuid, ok := d.FilesystemLabelToPartUUID[label]; ok {
+		return partuuid, nil
+	}
+	return "", fmt.Errorf("could not find label %q", label)
+}
+
+// HasPartitions returns if the mock disk has partitions or not. Part of the
+// Disk interface.
+func (d *MockDiskMapping) HasPartitions() bool {
+	panicIfNotTest()
+	return d.DiskHasPartitions
+}
+
+// MountPointIsFromDisk returns if the disk that the specified mount point comes
+// from is the same disk as the object. Part of the Disk interface.
+func (d *MockDiskMapping) MountPointIsFromDisk(mountpoint string, opts *Options) (bool, error) {
+	panicIfNotTest()
+
+	// this is relying on the fact that DiskFromMountPoint should have been
+	// mocked for us to be using this mockDisk method anyways
+	otherDisk, err := DiskFromMountPoint(mountpoint, opts)
+	if err != nil {
+		return false, err
+	}
+
+	if otherDisk.Dev() == d.Dev() && otherDisk.HasPartitions() == d.HasPartitions() {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// Dev returns a unique representation of the mock disk, it is a hash of the
+// mock disk struct string representation. Part of the Disk interface.
+func (d *MockDiskMapping) Dev() string {
+	panicIfNotTest()
```

Comment:
this feels a bit overcomplicated, but maybe I'm missing the final usage. Wouldn't it be simpler to have  a `Dev string` field in MockDiskMapping ?

Reply:
IMHO it's not overly complicated, but perhaps that's because I'm attached to the implementation because I already wrote it :smile: 

but you are right, that it would be simpler to just require that callers who are mocking disks to put a unique string inside the `MockDiskMapping` when they call MockMountPointDisksToPartionMapping. I will open a PR soon with the final usage here, but that PR will depend on the systemd-mount PR so it will be big, but hopefully it will be clear how it is used.

Reply:
perhaps to your point though I did actually find a bug with my initial implementation here and had to first hash the key and values separately rather than my naive writing of both the key and value to the hash.

Reply:
an example of how this package is currently used is on line 1751 of cmd_initramfs_mounts_test.go: https://github.com/snapcore/snapd/pull/9081/commits/79f6404f1ac044212cc984ae2415934a93cf9d7c

Reply:
Being attached to implementations will not bring you joy :)

a) this is a bit complicated
b) it produces a string that if it shows up in a test failure might be actually hard to correlate back to the test source of it

I recommend either to:
a) add the Dev field
b) OR use something like json.Marshal to produce an "ID" string  (json Marshal sorts object keys for you), that is easier to correlate back to test source

Reply:
this is refactored now to be simpler